### DISCUSSION TITLE
New deployments to Environment under the same branch no longer set the previous deployment as 'inactive' automatically

### DISCUSSION THREAD

**Discussion by JonathanAtCenterEdge** [community/community#67982] [Actions] [answered]
### Select Topic Area

Bug

### Body

Before two weeks ago every deployment to our `Staging - QA` environment would cause the previous deployment be set to a state of 'inactive' automatically, but starting two weeks ago every single new deployment is now permanently considered 'active' unless we manually set them to 'inactive' via the GitHub API when the workflow finishes - We previously didn't need to do this

![Wt31h9Uxd8](https://github.com/community/community/assets/91483632/04b153e9-148b-4bf3-b215-bb6a1d1bb123)

All of our deployments to `Staging - QA` fall under our `develop` branch, and if I'm understanding how deployments work correctly, any new deployments should set the previous deployment under the same environment and branch to 'inactive'

I'm not sure if this is a bug with my workflows or a bug with GitHub Actions

**Comment by SumanthSShet** [comment]
The described behavior is a change in the way that GitHub Actions deployments work. Starting in September 2023, GitHub Actions will no longer automatically set previous deployments to inactive when a new deployment is created to the same environment and branch. This is to give users more control over their deployments and to avoid inadvertently making previous deployments inactive.

To set a previous deployment to inactive, you can use the GitHub API. You can also use a GitHub Actions workflow to automate the process.

Here is an example of a GitHub Actions workflow that sets the previous deployment to inactive:

YAML
name: Set previous deployment to inactive

on:
  push:
    branches:
      - develop

jobs:
  set-previous-deployment-to-inactive:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/github-script@v6

        with:
          script: |
            // Get the previous deployment
            const previousDeployment = github.repos.getDeployment({
              owner: context.repo.owner,
              repo: context.repo.repo,
              deployment_id: context.payload.deployment.previous_deployment_id,
            });

            // Set the previous deployment to inactive
            await github.repos.updateDeploymentStatus({
              owner: context.repo.owner,
              repo: context.repo.repo,
              deployment_id: previousDeployment.id,
              state: 'inactive',
            });

This workflow will run whenever a push is made to the develop branch. It will use the GitHub API to get the previous deployment for the Staging - QA environment. Then, it will set the previous deployment to inactive.

You can add this workflow to your repository to automate the process of setting previous deployments to inactive.

**Reply by JonathanAtCenterEdge** [reply]
Was there an announcement that I missed about a change to this behavior? I'd like to know so that I could subscribe and know about it sooner

**Reply by SumanthSShet** [reply]
Yes, GitHub announced the change to the deployment status behavior in a blog post on August 4, 2023. You can subscribe to GitHub's blog to be notified of future announcements.

**Reply by JonathanAtCenterEdge** [reply]
Do you happen to have a link to that blog post? I'm having trouble finding it

**Reply by SumanthSShet** [reply]
Sure, here is a link to the GitHub blog post that announces the change to the deployment status behavior: https://docs.github.com/en/rest/deployments/statuses

**Reply by JonathanAtCenterEdge** [reply]
I don't see a blog post in that link, that appears to be a link to the REST API docs, which I don't see a notice of behavior anywhere.

**Comment by JamesMGreene** [comment]
I'm not aware of any recent changes to the auto-inactivating behavior of Deployments.

However, as Deployments is one of our oldest REST APIs and data models, the auto-inactivation has accrued some arguably quirky behaviors over its lifetime that could be affecting you if you have changed anything about your deployment process lately. 😅 

Are these deployments being created internally with the [GitHub Actions `environment:` configuration syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idenvironment)? That's the impression I'm getting from your initial post, but you didn't explicitly say that. If so, the most likely culprit is that you added or changed the environment URL specified in the workflow definition. 👀 

When creating Deployments internally using GitHub Actions workflows and the [per-job `environment:` configuration syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idenvironment), the rules and pitfalls are slightly different than the rules of the REST API. Auto-inactivation will only occur if **all** of the following are true:
- [ ] The new Deployment must complete via a final DeploymentStatus with `state: "success"`, i.e. the Actions workflow job run must end with a conclusion of `"success"`.
- [ ] The previous Deployments completed via a final DeploymentStatus with `state: "success"`. Doesn't matter if they were created internally with GitHub Actions or via the REST API.
- [ ] The previous Deployments have an environment name matching the environment name of the new successful Deployment.
- [ ] The previous Deployments were not marked as `transient_environment: true` when created. This is only possible if they were created via the REST API, not internally with GitHub Actions.
- [ ] ℹ️ The previous Deployments may have been marked as `production_environment: true` when created, and/or have the environment name of `production`. They will still be auto-inactivated by Actions. This behavior is a divergence from the REST API's default behavior.
- [ ] :warning: If the new Deployment specifies an environment URL in the workflow definition, only previous Deployments with the _exact same environment URL_ will be auto-inactivated. This behavior is a divergence from the REST API behavior.
  - ℹ️ Note: If the new Deployment does **not** specify an environment URL in the workflow definition, its auto-inactivation will still include previous Deployments that _did_ have an environment URL.

If these rules don't get your issue sorted, please reach out again with more information about your particular configuration. Alternatively, if you can reproduce this in a public repository, I'd also be happy to investigate further there.

Hopefully this helps! 🤞🏻

**Reply by JonathanAtCenterEdge** [reply]
First off, thank you for such an informative reply! You are correct, they are being created internally with the `environment: ` syntax, however we don't configure a URL, this is something I can try however. I can confirm that we haven't changed anything related to the `environment:` syntax in the past two weeks, the only time we touch the deployments via the REST API is in a completely different environment and different workflow.

```yaml
jobs:
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: read # Needed for actions/checkout
      id-token: write # Needed for requesting the JWT
      actions: read # Needed for downloading workflow artifacts
    environment:
      name: Staging - QA [stagedev]
    steps:
```

Is how I've declared our staging deployment workflow, this specific configuration hasn't changed since the workflows creation. The only modification to the workflow in the last two weeks, which notably when these changes were merged is when I noticed our deployments were no longer being auto-inactive'ed, have been switching off of some "local-action" `using` calls, for example `uses: ./.github/workflows/actions/kube-apply-deployment` to a centralized action inside a repository in our organization, `centeredge/cd-actions/kube/apply-deployment@v1`. From my perspective such changes shouldn't have caused it though.

Going through your list, I don't see why our workflow, as declared above, shouldn't be auto-inactivating.

- The deployment does complete with a final success
- The previous deployments do also complete with a final success, looking at the workflow runs 
- The previous deployment workflow run use a static environment name, so it isn't changing
- I'm using the internal `environment: ` syntax, so this rule doesn't apply
- While our Production environment does have Production in the name, it isn't explicitly `production`
- We don't define an environment URL at all

This is a much broader view of our deployment history for our staging environment
![image](https://github.com/community/community/assets/91483632/5edaca31-b79c-4895-822e-9f0e239a41a6)

**Reply by JonathanAtCenterEdge** [reply]
Hi @JamesMGreene, I've been able to replicate this in a public sandbox repository with minimal workflows https://github.com/JonathanAtCenterEdge/actions-environment-sandbox, you can see the deployment history here https://github.com/JonathanAtCenterEdge/actions-environment-sandbox/deployments

I can confirm that auto-inactive is not working with this simple GraphQL query
```graphql
query recentDeployment($owner: String!, $repo: String!, $environment: String!) {
  repository(owner:$owner,name:$repo) {
    deployments(first: 10, environments: [$environment], orderBy: { field: CREATED_AT, direction: DESC } ) {
      nodes {
        state
        createdAt
      }
    }
  }
}
```
With variables `{"owner": "JonathanAtCenterEdge", "repo": "actions-environment-sandbox", "environment": "Staging"} `

Which gives me
```json
{
  "data": {
    "repository": {
      "deployments": {
        "nodes": [
          {
            "state": "ACTIVE",
            "createdAt": "2023-09-25T17:51:49Z",
            "environment": "Staging",
            "databaseId": 1090684465,
            "commit": {
              "oid": "07c9f5d247cec0b662453d7c121d4483c8f19448"
            }
          },
          {
            "state": "ACTIVE",
            "createdAt": "2023-09-25T17:51:07Z",
            "environment": "Staging",
            "databaseId": 1090683372,
            "commit": {
              "oid": "07c9f5d247cec0b662453d7c121d4483c8f19448"
            }
          }
        ]
      }
    }
  }
}
```

EDIT: Further testing with using `url:` the field under `environment` reveals that auto-inactive does work for deployments that have the same URL as a previous deployment, it looks like deployments with blank URLs' might be currently bugged

EDIT 2: Hm, it looks like setting the environment url did not actually do it in our repository

**Reply by JonathanAtCenterEdge** [reply]
Two weeks later this still isn't functioning correctly as seen in this repository https://github.com/JonathanAtCenterEdge/actions-environment-sandbox

**Reply by JamesMGreene** [reply]
@JonathanAtCenterEdge Any idea what happened with these few that did end up marked as inactive? 🤔 

<img width="1269" alt="some inactive deployments" src="https://github.com/community/community/assets/417751/a38852a7-5107-463c-b79d-4f8ea47a1b07">

**Reply by JonathanAtCenterEdge** [reply]
> @JonathanAtCenterEdge Any idea what happened with these few that did end up marked as inactive? 🤔 
> 
> <img width="1269" alt="some inactive deployments" src="https://github.com/community/community/assets/417751/a38852a7-5107-463c-b79d-4f8ea47a1b07">
> 

Those had URLs applied to them, I added them, ran a few deployments which tricked the auto inactive to kick in, then removed the URLs later

EDIT: My mistake, I think I modified the URLs in those deployments, I think I was using the same URL as the production environment https://github.com/JonathanAtCenterEdge/actions-environment-sandbox/deployments/Staging This environment might have a clearer picture with and without URLs

**Reply by JamesMGreene** [reply]
I see a commit to GitHub from September 13 that may have affected this. Would that fit into your timeline for when you began noticing this issue?

**Reply by JonathanAtCenterEdge** [reply]
> I see a commit to GitHub from September 13 that may have affected this. Would that fit into your timeline for when you began noticing this issue?

It does, all deployments on or after September 14th are affected for us.

**Reply by JamesMGreene** [reply]
Thanks for confirming. I've added a bug to our backlog for this, we'll try to get to it sooner than later. Sorry for the inconvenience in the meanwhile. 😓

**Comment by JamesMGreene** [comment, marked as answer]
@JonathanAtCenterEdge Checking back in! 👋🏻  We believe this should be resolved as of ~18 hours ago. Please take a look when you get a chance, and let us know if you're still seeing issues with the auto-inactivating behavior going forward. Thanks again for the report and your patience! ❤️
